common/src/block_interval.rs

```diff
@@ -0,0 +1,294 @@
+    pub fn into_bounded_stream(self) -> Result<impl Stream<Item = u64>, BlockIntervalError> {
+        match self {
+            BlockInterval::SingleBlockId(BlockId::Number(num)) => {
+                let num = num
+                    .as_number()
+                    .ok_or(BlockIntervalError::InvalidBlockNumber(num.to_string()))?;
+                Ok(tokio_stream::iter(num..num + 1))
+            }
+            BlockInterval::Range(range) => Ok(tokio_stream::iter(range)),
+            _ => Err(BlockIntervalError::IntoBoundedStreamError),
+        }
+    }
+
+    /// Convert the block interval into an unbounded async stream of block
+    /// numbers. Query the blockchain node for the latest block number.
+    pub async fn into_unbounded_stream<ProviderT, TransportT>(
+        self,
+        provider: ProviderT,
+    ) -> Result<impl Stream<Item = Result<u64, anyhow::Error>>, BlockIntervalError>
+    where
+        ProviderT: Provider<TransportT>,
+        TransportT: Transport + Clone,
+    {
+        match self {
+            BlockInterval::FollowFrom {
+                start_block,
+                block_time,
+            } => Ok(try_stream! {
+                let mut current = start_block;
+                 loop {
+                    let last_block_number = provider.get_block_number().await.map_err(|e: alloy::transports::RpcError<_>| {
+                        error!("Unable to retrieve latest chain block number {e}");
+                        BlockIntervalError::LatestBlockNumberError(e.to_string())
+                    })?;
+
+                    if current < last_block_number {
+                        current += 1;
+                        yield current;
+                    } else {
+                       info!("Waiting for the new blocks to be mined, requested block number: {current}, \
+                       latest block number: {last_block_number}");
+                        let block_time = block_time.unwrap_or(DEFAULT_BLOCK_TIME);
+                        // No need to poll the node too frequently, waiting
+                        // a block time interval for a block to be mined should be enough
+                       tokio::time::sleep(tokio::time::Duration::from_millis(block_time)).await;
+                    }
+                }
+            }),
+            _ => Err(BlockIntervalError::IntoUnboundedStreamError),
+        }
+    }
+}
+
+impl std::fmt::Display for BlockInterval {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            BlockInterval::SingleBlockId(block_id) => {
+                write!(
+                    f,
+                    "{}",
+                    match block_id {
+                        BlockId::Hash(rpc_hash) =>
+                            "0x".to_string() + &hex::encode(rpc_hash.block_hash),
+                        BlockId::Number(number) => number.to_string(),
+                    }
+                )
```

Comment:
```suggestion
match block_id {
    BlockId::Number(it) => f.write_fmt(format_args!("{}", it)),
    BlockId::Hash(it) => f.write_fmt(format_args!("0x{}", it)),
}
```

No need to be that verbose or allocate strings

Reply:
```
error[E0277]: `alloy::rpc::types::alloy_rpc_types::RpcBlockHash` doesn't implement `std::fmt::Display`
`alloy::rpc::types::alloy_rpc_types::RpcBlockHash` cannot be formatted with the default formatter
```

Reply:
```
BlockId::Hash(it) => f.write_fmt(format_args!("0x{}", &hex::encode(it.block_hash))),
```

Reply:
I'm not at my laptop this minute, but use the fmt::Hex format specifier, check the docs on std::fmt

Reply:
It is not specified for the byte array. One of the main reasons `hex` crate exists.